### PR TITLE
DAOS-3783 rdb: use epoch in record for installing snapshot

### PR DESCRIPTION
**Pull request by wangdi1**
Since epoch is not allowed to be 0 after new punch model,
Let's use epoch in record for snapshot replication.

Signed-off-by: Di Wang <di.wang@intel.com>

### CODE REVIEW

**Comment by liw** [comment]
@jolivier23, how should we update vos_obj_update's documentation for the epoch parameter?

**Comment by jolivier23** [comment]
> @jolivier23, how should we update vos_obj_update's documentation for the epoch parameter?

Indeed, it should be updated.   VOS ignores the epoch ranges now entirely.  I would like to remove them from the IOD struct entirely.   They were previously needed for rebuild but now we do logical rebuild at a snapshot.   Is this an issue for RDB use case?

**Comment by liw** [comment]
> @liw Indeed, it should be updated. VOS ignores the epoch ranges now entirely. I would like to remove them from the IOD struct entirely. They were previously needed for rebuild but now we do logical rebuild at a snapshot. Is this an issue for RDB use case?

@jolivier23, does not sounds like an issue. So during dss_enum_unpack, rdb should just pass its snapshot epoch to vos_obj_update, disregarding the epoch ranges?

**Comment by jolivier23** [comment]
> > @liw Indeed, it should be updated. VOS ignores the epoch ranges now entirely. I would like to remove them from the IOD struct entirely. They were previously needed for rebuild but now we do logical rebuild at a snapshot. Is this an issue for RDB use case?
> 
> @jolivier23, does not sounds like an issue. So during dss_enum_unpack, rdb should just pass its snapshot epoch to vos_obj_update, disregarding the epoch ranges?

@liw yes, that would be ideal.   One caveat is that it does require separate calls if dealing with multiple snapshots.

File: src/rdb/rdb_raft.c [hunk 1]
```diff
@@ -704,8 +704,21 @@ rdb_raft_exec_unpack_io(struct dss_enum_unpack_io *io, void *arg)
 		D_ASSERT(io->ui_sgls[i].sg_iovs[0].iov_len > 0);
 	}
 #endif
+	daos_epoch_t epoch = 0;
+	int i;
+
+	for (i = 0; i <= io->ui_iods_top; i++) {
```
Comment (jolivier23) [inline src/rdb/rdb_raft.c]: In lieu of the discussion on this, is there a better way to do this?

Comment (wangdi1) [reply src/rdb/rdb_raft.c]: ok, then I will use input snapshot idx as the epoch.

**Comment by wangdi1** [comment]
close this one, Liwei will resolve it by PR-1499

**Pull request merged**